### PR TITLE
Add arrays comparison in QgsExpression

### PR DESCRIPTION
**Pull request by arnaud-morvan**
## Description
This adds the arrays comparison operators in QgsExpression. I've used PostgreSQL behavior as a model. Before this arrays comparion of QVariantList objects was made on a string convertion using QVariantList.toString() => empty string "" (always equal).

Note that I do not have tested nor handled other operators (plus, minus, ...).

I've also added QList objects (arrays) in QgsExpression::quotedValue. Before this  QgsExpression::quotedValue( QVariantList ) was always returning empty string : "".

For instance, this give good results in aggregate functions with arrays (two fields) in group_by parameter.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/core/qgsexpression.cpp [hunk 1]
```diff
@@ -4803,6 +4818,75 @@ QVariant QgsExpression::NodeBinaryOperator::eval( QgsExpression *parent, const Q
...
+        // verify that we have two lists
+        if ( !isList( vL ) || !isList( vR ) )
+          return TVL_Unknown;
+
+        // and search for not equal respective items
+        QVariantList lL = vL.toList();
+        QVariantList lR = vR.toList();
+        for ( int i = 0; i < lL.length() && i < lR.length(); i++ )
+        {
+          if ( isNull( lL[i] ) && isNull( lR[i] ) )
```
Comment (m-kuhn) [inline src/core/qgsexpression.cpp]: Better use `iL.at(i)` to prevent a detach.

File: src/core/qgsexpression.cpp [hunk 2]
```diff
@@ -4803,6 +4818,75 @@ QVariant QgsExpression::NodeBinaryOperator::eval( QgsExpression *parent, const Q
...
+              case boGE:
+                return isNull( lL[i] );
+              default:
+                Q_ASSERT( false );
+                return TVL_Unknown;
+            }
+          }
+
+          QgsExpression::NodeLiteral nL( lL[i] );
+          QgsExpression::NodeLiteral nR( lR[i] );
```
Comment (m-kuhn) [inline src/core/qgsexpression.cpp]: .at

**Comment by m-kuhn** [comment]
Nice

File: src/core/qgsexpression.cpp [hunk 3]
```diff
@@ -4225,6 +4230,16 @@ QString QgsExpression::quotedValue( const QVariant &value, QVariant::Type type )
...
       return value.toBool() ? "TRUE" : "FALSE";
 
+    case QVariant::List:
+    {
+      QStringList quotedValues;
+      Q_FOREACH ( const QVariant &v, value.toList() )
+      {
+        quotedValues += quotedValue( v );
+      }
+      return QStringLiteral( "array( %1 )" ).arg( quotedValues.join( QStringLiteral( ", " ) ) );
```
Comment (nyalldawson) [inline src/core/qgsexpression.cpp]: If you change this to use the char based join it'll be MUCH faster:

    return QStringLiteral( "array( %1 )" ).arg( quotedValues.join( ',' ) ) );

File: tests/src/core/testqgsexpression.cpp [hunk 4]
```diff
@@ -2443,6 +2443,43 @@ class TestQgsExpression: public QObject
       QCOMPARE( badArray.evalErrorString(), QString( "Cannot convert 'not an array' to array" ) );
     }
 
+    void compare_arrays()
+    {
+      QCOMPARE( QgsExpression( "array() = array()" ).evaluate(), QVariant( true ) );
+      QCOMPARE( QgsExpression( "array(NULL) = array(NULL)" ).evaluate(), QVariant( true ) );
```
Comment (nyalldawson) [inline tests/src/core/testqgsexpression.cpp]: This is odd - I'd have expected the result to be null here (to match null=null). But you're right, Postgres reports this as true, so leave it this way.

File: tests/src/core/testqgsexpression.cpp [hunk 5]
```diff
@@ -2443,6 +2443,43 @@ class TestQgsExpression: public QObject
...
+      QCOMPARE( QgsExpression( "array('hello') = array('hello2')" ).evaluate(), QVariant( false ) );
+
+      QCOMPARE( QgsExpression( "array('1') = array(1)" ).evaluate(), QVariant( true ) );
+      QCOMPARE( QgsExpression( "array('1.2') = array(1.2)" ).evaluate(), QVariant( true ) );
+
+      QCOMPARE( QgsExpression( "array() != array()" ).evaluate(), QVariant( false ) );
+      QCOMPARE( QgsExpression( "array(NULL) != array(NULL)" ).evaluate(), QVariant( false ) );
+      QCOMPARE( QgsExpression( "array() != array(NULL)" ).evaluate(), QVariant( true ) );
+      QCOMPARE( QgsExpression( "array('hello') != array('hello')" ).evaluate(), QVariant( false ) );
+      QCOMPARE( QgsExpression( "array('hello') != array('hello2')" ).evaluate(), QVariant( true ) );
```
Comment (nyalldawson) [inline tests/src/core/testqgsexpression.cpp]: I don't see any equality tests here with multiple array values - I think they should be added too. Also tests with differing array lengths.

**Review by nyalldawson — COMMENTED** [review]
Great changes!

File: tests/src/core/testqgsexpression.cpp (hunk 4 above)
Comment (m-kuhn) [inline tests/src/core/testqgsexpression.cpp]: That's what I've learnt in oracle DB courses 15 years ago: `NULL` is not equal to `NULL`.

Comment (arnaud-morvan) [inline tests/src/core/testqgsexpression.cpp]: I agree, but regarding PostgreSQL ( Array[NULL] = Array[NULL] ) => true.
For me this make sense, for instance if you want to make a group_by on two columns.
Array[1, NULL] = Array[1, NULL] => True.
Without this you will lost some records.

**Comment by arnaud-morvan** [comment]
I've added some tests with multiple array values and differing array length, for me this is ready for merging.

**Comment by arnaud-morvan** [comment]
@nyalldawson @m-kuhn : Thanks for review, I've added some tests, for me this is ready for merging now, but I do not yet have push rights.

**Comment by m-kuhn** [comment]
LGTM, @nyalldawson, are you still waiting for something?

**Comment by m-kuhn** [comment]
Yay, thanks @arnaud-morvan

**Pull request merged**